6-1. 프로젝트 (1) load_digits : 손글씨를 분류해 봅시다

(1) 필요한 모듈 import 하기

In [10]:
import sklearn

print(sklearn.__version__)

1.5.1


In [1]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

(2) 데이터 준비

In [2]:
dg = load_digits()

(3) 데이터 이해하기

In [3]:
dg_data = dg.data # Feature Data 지정하기
dg_label = dg.target # Label Data 지정하기

print(dg.target_names) # Target Names 출력해보기

[0 1 2 3 4 5 6 7 8 9]


In [4]:
print(dg.DESCR) # 데이터 Describe 해 보기

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

:Number of Instances: 1797
:Number of Attributes: 64
:Attribute Information: 8x8 image of integer pixels in the range 0..16.
:Missing Attribute Values: None
:Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
:Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each block. This generates
an in

(4) train, test 데이터 분리 모델 학습과 테스트용 문제지와 정답지를 준비해 봅시다.

In [5]:
X_train, X_test, y_train, y_test = train_test_split(dg_data,
                                                    dg_label,
                                                    test_size=0.2,
                                                    random_state=0)

(5) 다양한 모델로 학습시켜보기

In [6]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

In [7]:
# Decision Tree 사용해보기
dg_dt = DecisionTreeClassifier(random_state=0)
dg_dt.fit(X_train, y_train)
y_pred = dg_dt.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.89      0.87        27
           1       0.84      0.89      0.86        35
           2       0.78      0.81      0.79        36
           3       0.69      0.93      0.79        29
           4       0.93      0.87      0.90        30
           5       0.92      0.85      0.88        40
           6       0.87      0.89      0.88        44
           7       0.90      0.95      0.93        39
           8       0.89      0.62      0.73        39
           9       0.83      0.83      0.83        41

    accuracy                           0.85       360
   macro avg       0.85      0.85      0.85       360
weighted avg       0.85      0.85      0.85       360



In [8]:
# Random Forest 사용해 보기
dg_rf = RandomForestClassifier(random_state=0)
dg_rf.fit(X_train, y_train)
y_pred = dg_rf.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98        27
           1       0.94      0.97      0.96        35
           2       1.00      0.94      0.97        36
           3       0.94      1.00      0.97        29
           4       1.00      0.97      0.98        30
           5       0.93      0.95      0.94        40
           6       1.00      1.00      1.00        44
           7       0.97      1.00      0.99        39
           8       1.00      0.97      0.99        39
           9       0.95      0.90      0.93        41

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



In [12]:
# SVM 사용해 보기
dg_svm = svm.SVC(random_state=0)
dg_svm.fit(X_train, y_train)
y_pred = dg_svm.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        27
           1       0.97      1.00      0.99        35
           2       1.00      1.00      1.00        36
           3       1.00      1.00      1.00        29
           4       1.00      1.00      1.00        30
           5       0.97      0.97      0.97        40
           6       1.00      1.00      1.00        44
           7       1.00      1.00      1.00        39
           8       1.00      0.97      0.99        39
           9       0.98      0.98      0.98        41

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [13]:
# SGD Classifier 사용해 보기
dg_sgd = SGDClassifier(random_state=0)
dg_sgd.fit(X_train, y_train)
y_pred = dg_sgd.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        27
           1       0.86      0.91      0.89        35
           2       1.00      0.94      0.97        36
           3       0.97      0.97      0.97        29
           4       0.97      0.97      0.97        30
           5       0.97      0.93      0.95        40
           6       0.98      0.98      0.98        44
           7       0.97      0.95      0.96        39
           8       0.78      0.97      0.86        39
           9       0.97      0.80      0.88        41

    accuracy                           0.94       360
   macro avg       0.95      0.94      0.94       360
weighted avg       0.95      0.94      0.94       360



In [14]:
# Logistic Regression 사용해 보기
dg_lr = LogisticRegression(random_state=0)
dg_lr.fit(X_train, y_train)
y_pred = dg_lr.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        27
           1       0.92      0.97      0.94        35
           2       0.97      0.97      0.97        36
           3       0.97      1.00      0.98        29
           4       0.97      0.97      0.97        30
           5       0.97      0.93      0.95        40
           6       1.00      0.98      0.99        44
           7       0.97      0.97      0.97        39
           8       0.97      0.92      0.95        39
           9       0.93      0.98      0.95        41

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



c:\Users\user\miniconda3\envs\aifenv\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(6) 모델을 평가해 보기

In [15]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_test, y_pred))

[[27  0  0  0  0  0  0  0  0  0]
 [ 0 34  0  0  0  0  0  0  1  0]
 [ 0  0 35  1  0  0  0  0  0  0]
 [ 0  0  0 29  0  0  0  0  0  0]
 [ 0  0  0  0 29  0  0  1  0  0]
 [ 0  0  0  0  0 37  0  0  0  3]
 [ 0  1  0  0  0  0 43  0  0  0]
 [ 0  0  0  0  1  0  0 38  0  0]
 [ 0  2  1  0  0  0  0  0 36  0]
 [ 0  0  0  0  0  1  0  0  0 40]]


어떤 숫자의 글씨가 잘못 분류되었는지 한 눈에 확인할 수 있기에 혼동행렬로 평가지표로 활용하였습니다. Logistic Regression 모델을 통해 혼동행렬을 보았을 때 8을 예측했을 때 가장 잘못 분류됨을 알 수 있습니다.

6-2. 프로젝트 (2) load_wine : 와인을 분류해 봅시다

(1) 필요한 모듈 import 하기

In [16]:
from sklearn.datasets import load_wine

(2) 데이터 준비

In [17]:
wine = load_wine()

(3) 데이터 이해하기

In [18]:
wine_data = wine.data # Feature Data 지정하기
wine_label = wine.target # Label Data 지정하기

print(wine.target_names) # Target Names 출력해보기

['class_0' 'class_1' 'class_2']


In [19]:
print(wine.DESCR) # Describe 해 보기

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

:Number of Instances: 178
:Number of Attributes: 13 numeric, predictive attributes and the class
:Attribute Information:
    - Alcohol
    - Malic acid
    - Ash
    - Alcalinity of ash
    - Magnesium
    - Total phenols
    - Flavanoids
    - Nonflavanoid phenols
    - Proanthocyanins
    - Color intensity
    - Hue
    - OD280/OD315 of diluted wines
    - Proline
    - class:
        - class_0
        - class_1
        - class_2

:Summary Statistics:

============================= ==== ===== ======= =====
                                Min   Max   Mean     SD
============================= ==== ===== ======= =====
Alcohol:                      11.0  14.8    13.0   0.8
Malic Acid:                   0.74  5.80    2.34  1.12
Ash:                          1.36  3.23    2.36  0.27
Alcalinity of Ash:            10.6  30.0    19.5   3.3
Magnesium:                    70.0 162.0    99.7  14.3

(4) train, test 데이터 분리

In [20]:
X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_label,
                                                    test_size=0.2,
                                                    random_state=2)

(5) 다양한 모델로 학습시켜보기

In [21]:
# Decision Tree 사용해 보기기
wine_dt = DecisionTreeClassifier(random_state=2)
wine_dt.fit(X_train, y_train)
y_pred = wine_dt.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.89      0.94        18
           1       0.82      1.00      0.90         9
           2       1.00      1.00      1.00         9

    accuracy                           0.94        36
   macro avg       0.94      0.96      0.95        36
weighted avg       0.95      0.94      0.95        36



In [22]:
# Random Forest 사용해 보기
wine_rf = RandomForestClassifier(random_state=2)
wine_rf.fit(X_train, y_train)
y_pred = wine_rf.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        18
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00         9

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



In [23]:
# SVM 사용해 보기
wine_svm = svm.SVC(random_state=2)
wine_svm.fit(X_train, y_train)
y_pred = wine_svm.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.89      0.91        18
           1       0.53      1.00      0.69         9
           2       0.00      0.00      0.00         9

    accuracy                           0.69        36
   macro avg       0.49      0.63      0.54        36
weighted avg       0.60      0.69      0.63        36



In [24]:
# SGD Classifier 사용해 보기
wine_sgd = SGDClassifier(random_state=2)
wine_sgd.fit(X_train, y_train)
y_pred = wine_sgd.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.83      0.91        18
           1       0.62      0.89      0.73         9
           2       0.50      0.44      0.47         9

    accuracy                           0.75        36
   macro avg       0.71      0.72      0.70        36
weighted avg       0.78      0.75      0.75        36



In [25]:
# Logistic Regression 사용해 보기
wine_lr = LogisticRegression(random_state=2)
wine_lr.fit(X_train, y_train)
y_pred = wine_lr.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.89      0.94        18
           1       0.75      1.00      0.86         9
           2       1.00      0.89      0.94         9

    accuracy                           0.92        36
   macro avg       0.92      0.93      0.91        36
weighted avg       0.94      0.92      0.92        36



c:\Users\user\miniconda3\envs\aifenv\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(6) 모델을 평가해 보기

In [26]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.89      0.94        18
           1       0.75      1.00      0.86         9
           2       1.00      0.89      0.94         9

    accuracy                           0.92        36
   macro avg       0.92      0.93      0.91        36
weighted avg       0.94      0.92      0.92        36



F1-score을 평가 지표로 활용하는것은 정밀도와 재현율의 조화평균으로 저급 와인을 고급 와인으로, 고급 와인을 저급 와인으로 잘못 분류하는 것 모두 발생 가능하기에, 정밀도와 재현율을 모두 고려하기 위함 현재 지표에서는 class_1가 가장 잘못 분류되고 있음을 알 수 있다.

6-3. 프로젝트(3) load_breast_cancer : 유방암 여부를 진단해 봅시다

(1) 필요한 모듈 import 하기

In [27]:
from sklearn.datasets import load_breast_cancer

(2) 데이터 준비 

In [28]:
bc = load_breast_cancer()

(3) 데이터 이해하기

In [29]:
bc_data = bc.data # Feature Data 지정하기
bc_label = bc.target # Label Data 지정하기

print(bc.target_names) # Target Names 출력해 보기

['malignant' 'benign']


In [30]:
print(bc.DESCR) # 데이터 Describe 해 보기

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

:Number of Instances: 569

:Number of Attributes: 30 numeric, predictive attributes and the class

:Attribute Information:
    - radius (mean of distances from center to points on the perimeter)
    - texture (standard deviation of gray-scale values)
    - perimeter
    - area
    - smoothness (local variation in radius lengths)
    - compactness (perimeter^2 / area - 1.0)
    - concavity (severity of concave portions of the contour)
    - concave points (number of concave portions of the contour)
    - symmetry
    - fractal dimension ("coastline approximation" - 1)

    The mean, standard error, and "worst" or largest (mean of the three
    worst/largest values) of these features were computed for each image,
    resulting in 30 features.  For instance, field 0 is Mean Radius, field
    10 is Radius SE, field 20 is Worst Radius.

    - 

(4) train, test 데이터 분리

In [31]:
X_train, X_test, y_train, y_test = train_test_split(bc_data,
                                                    bc_label,
                                                    test_size=0.2,
                                                    random_state=4)

(5) 다양한 모델로 학습시켜보기

In [32]:
# Decision Tree 사용해 보기기
bc_dt = DecisionTreeClassifier(random_state=4)
bc_dt.fit(X_train, y_train)
y_pred = bc_dt.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.74      1.00      0.85        34
           1       1.00      0.85      0.92        80

    accuracy                           0.89       114
   macro avg       0.87      0.93      0.88       114
weighted avg       0.92      0.89      0.90       114



In [33]:
# Random Forest 사용해 보기
bc_rf = RandomForestClassifier(random_state=4)
bc_rf.fit(X_train, y_train)
y_pred = bc_rf.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.79      0.97      0.87        34
           1       0.99      0.89      0.93        80

    accuracy                           0.91       114
   macro avg       0.89      0.93      0.90       114
weighted avg       0.93      0.91      0.91       114



In [34]:
# SVM 사용해 보기
bc_svm = svm.SVC(random_state=4)
bc_svm.fit(X_train, y_train)
y_pred = bc_svm.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.76      0.82      0.79        34
           1       0.92      0.89      0.90        80

    accuracy                           0.87       114
   macro avg       0.84      0.86      0.85       114
weighted avg       0.87      0.87      0.87       114



In [35]:
# SGD Classifier 사용해 보기
bc_sgd = SGDClassifier(random_state=4)
bc_sgd.fit(X_train, y_train)
y_pred = bc_sgd.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.61      0.88      0.72        34
           1       0.94      0.76      0.84        80

    accuracy                           0.80       114
   macro avg       0.78      0.82      0.78       114
weighted avg       0.84      0.80      0.81       114



In [36]:
# Logistic Regression 사용해 보기
bc_lr = LogisticRegression(random_state=4)
bc_lr.fit(X_train, y_train)
y_pred = bc_lr.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.78      0.85      0.82        34
           1       0.94      0.90      0.92        80

    accuracy                           0.89       114
   macro avg       0.86      0.88      0.87       114
weighted avg       0.89      0.89      0.89       114



c:\Users\user\miniconda3\envs\aifenv\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(6) 모델을 평가해 보기

In [37]:
from sklearn.metrics import recall_score

print(recall_score(y_test, y_pred))

0.9


유방암 진단을 할 때, 음성을 양성으로 진단하는 것보다 양성을 음성으로 진단하는 것이 큰 문제이기때문에 재현율을 평가하였고 0.9의 재현율을 보이니 114명의 환자 중 약 11명의 환자들이 양성을 음성으로 진단받고있음을 알 수 있다.